File: DSAinJava/BinaryTree/FlipEquvilanceBinarytree.java
/**
 * leetcode problem no : (551)
 * 
 * leetcode solution link (https://leetcode.com/problems/flip-equivalent-binary-trees/)
 */


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    private boolean f(TreeNode root1, TreeNode root2){
        //BAse case 

        if(root1 == null && root2 == null){
            return true;
        }
        if(root1  == null || root2 == null) {
            // if any root is null or another contain value then , return false
            return false;
        }

        if(root1.val != root2.val ){
            return false;
        }else{
            return (f(root1.left , root2.left)&&f(root1.right , root2.right))
            ||(f(root1.left, root2.right) && f(root1.right, root2.left));
            
        }
    }

    public boolean flipEquiv(TreeNode root1, TreeNode root2) {

        return f(root1 , root2);
        
    }
}

File: DSAjavascript/BinarryTree/FlipEquvilanceBinarytree.js
/**
 * leetcode problem no : (551)
 * 
 * leetcode solution link (https://leetcode.com/problems/flip-equivalent-binary-trees/)
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
function f(root1 , root2){

    //Base casse 
    if(root1 == null && root2 == null)
    return true;

    if(root1 == null || root2 == null){
    // one of them if null
    return false;
    }

    if(root1.val != root2.val){

    return false;

    }else{
        return (f(root1.left ,root2.left )&& f(root1.right , root2.right))||(f(root1.left ,root2.right )&& f(root1.right , root2.left));
    }






 }
var flipEquiv = function(root1, root2) {
    

    return f(root1 , root2);
};